<a href="https://colab.research.google.com/github/NamithaDaney/LinearRegression/blob/main/Assignment_06_Linear_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**1. Loading and Preprocessing**


1.   Load the California Housing dataset using the fetch_california_housing function
from sklearn.
2.   Convert the dataset into a pandas DataFrame for easier handling. Handle
missing values (if any) and perform necessary feature scaling (e.g.,
standardization).
3. Explain the preprocessing steps you performed and justify why they are
necessary for this dataset.






In [188]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error,r2_score,mean_absolute_error,root_mean_squared_error,accuracy_score

In [189]:
# Load the California Housing dataset
data = fetch_california_housing()

# Convert the dataset into a pandas DataFrame
df = pd.DataFrame(data.data, columns=data.feature_names)
df['target'] = data.target
#df['MedHouseVal'] = california_housing.target'''

In [190]:
df

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847


In [191]:
data

{'data': array([[   8.3252    ,   41.        ,    6.98412698, ...,    2.55555556,
           37.88      , -122.23      ],
        [   8.3014    ,   21.        ,    6.23813708, ...,    2.10984183,
           37.86      , -122.22      ],
        [   7.2574    ,   52.        ,    8.28813559, ...,    2.80225989,
           37.85      , -122.24      ],
        ...,
        [   1.7       ,   17.        ,    5.20554273, ...,    2.3256351 ,
           39.43      , -121.22      ],
        [   1.8672    ,   18.        ,    5.32951289, ...,    2.12320917,
           39.43      , -121.32      ],
        [   2.3886    ,   16.        ,    5.25471698, ...,    2.61698113,
           39.37      , -121.24      ]]),
 'target': array([4.526, 3.585, 3.521, ..., 0.923, 0.847, 0.894]),
 'frame': None,
 'target_names': ['MedHouseVal'],
 'feature_names': ['MedInc',
  'HouseAge',
  'AveRooms',
  'AveBedrms',
  'Population',
  'AveOccup',
  'Latitude',
  'Longitude'],
 'DESCR': '.. _california_housing_dataset:\n

In [192]:
df

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422
...,...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09,0.781
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21,0.771
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22,0.923
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32,0.847


In [193]:
df.columns

Index(['MedInc', 'HouseAge', 'AveRooms', 'AveBedrms', 'Population', 'AveOccup',
       'Latitude', 'Longitude', 'target'],
      dtype='object')

In [194]:
df.isnull().sum()

,0
MedInc,0
HouseAge,0
AveRooms,0
AveBedrms,0
Population,0
AveOccup,0
Latitude,0
Longitude,0
target,0


In [195]:
df.duplicated().sum()

np.int64(0)

In [196]:
df.head()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,target
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


In [197]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Latitude    20640 non-null  float64
 7   Longitude   20640 non-null  float64
 8   target      20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB


In [198]:
df.describe()

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,target
count,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,3.870671,28.639486,5.429000,1.096675,1425.476744,3.070655,35.631861,-119.569704,2.068558
std,1.899822,12.585558,2.474173,0.473911,1132.462122,10.386050,2.135952,2.003532,1.153956
min,0.499900,1.000000,0.846154,0.333333,3.000000,0.692308,32.540000,-124.350000,0.149990
25%,2.563400,18.000000,4.440716,1.006079,787.000000,2.429741,33.930000,-121.800000,1.196000
50%,3.534800,29.000000,5.229129,1.048780,1166.000000,2.818116,34.260000,-118.490000,1.797000
75%,4.743250,37.000000,6.052381,1.099526,1725.000000,3.282261,37.710000,-118.010000,2.647250
max,15.000100,52.000000,141.909091,34.066667,35682.000000,1243.333333,41.950000,-114.310000,5.000010


In [199]:
df.nunique()

,0
MedInc,12928
HouseAge,52
AveRooms,19392
AveBedrms,14233
Population,3888
AveOccup,18841
Latitude,862
Longitude,844
target,3842


In [200]:
df.shape


(20640, 9)

Regression Algorithm Implementation (5 marks):
● Implement the following regression algorithms:

*   Linear Regression
*   Decision Tree Regressor
*   Random Forest Regressor
*   Gradient Boosting Regressor
*   Support Vector Regressor (SVR)

 For each algorithm: Provide a brief explanation of how it works. Explain why it might be suitable for this dataset.

In [201]:
X = df.drop('target', axis=1)
y = pd.Series(data.target)

In [202]:
X

,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25
...,...,...,...,...,...,...,...,...
20635,1.5603,25.0,5.045455,1.133333,845.0,2.560606,39.48,-121.09
20636,2.5568,18.0,6.114035,1.315789,356.0,3.122807,39.49,-121.21
20637,1.7000,17.0,5.205543,1.120092,1007.0,2.325635,39.43,-121.22
20638,1.8672,18.0,5.329513,1.171920,741.0,2.123209,39.43,-121.32


In [203]:
y

,0
0,4.526
1,3.585
2,3.521
3,3.413
4,3.422
...,...
20635,0.781
20636,0.771
20637,0.923
20638,0.847


In [204]:
x_train,x_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [205]:
#standardization
scalar = StandardScaler()
x_train_scaled = scalar.fit_transform(x_train)
x_test_scaled = scalar.transform(x_test)

We first check for null values and duplicate rows to ensure the dataset is clean and free of repetitive data that could bias our model. Finally, we standardize all features using a StandardScaler so that variables with large numbers, like population, do not mathematically overwhelm smaller numbers like income.

In [206]:
#Linear Regression
model = LinearRegression()
model.fit(x_train_scaled,y_train)

LinearRegression()

In [207]:
model.intercept_

np.float64(2.0719469373788777)

In [208]:
model.coef_

array([ 0.85438303,  0.12254624, -0.29441013,  0.33925949, -0.00230772,
       -0.0408291 , -0.89692888, -0.86984178])

In [209]:
y_predict = model.predict(x_test_scaled)
y_predict

array([0.71912284, 1.76401657, 2.70965883, ..., 4.46877017, 1.18751119,
       2.00940251])

In [210]:
mse = mean_squared_error(y_test, y_predict)
mse

0.5558915986952442

In [211]:
rmse = np.sqrt(mse)
rmse

np.float64(0.7455813830127763)

In [212]:
r2 = r2_score(y_test, y_predict)
r2

0.575787706032451

Linear regression works well here because California house prices strongly follow a simple rule: when neighborhood income goes up, home values go up. It is a fast, straightforward tool that tells you exactly how much each feature changes the price. This makes the results very easy to understand and explain to anyone.

In [213]:
#Decision Tree Regressor
from sklearn.tree import DecisionTreeRegressor, plot_tree

In [214]:
dt_model = DecisionTreeRegressor(criterion='squared_error', max_depth=5, random_state=42)
dt_model.fit(x_train_scaled, y_train)

DecisionTreeRegressor(max_depth=5, random_state=42)

In [215]:
y_dpredict = model.predict(x_test_scaled)
y_dpredict

array([0.71912284, 1.76401657, 2.70965883, ..., 4.46877017, 1.18751119,
       2.00940251])

In [216]:
#Accuracy
print('Mean Absolute Error:', mean_absolute_error(y_test, y_dpredict))
print('Mean Squared Error:', mean_squared_error(y_test, y_dpredict))
print('Root Mean Squared Error:', np.sqrt(mean_squared_error(y_test, y_dpredict)))
print('R2 Score:', r2_score(y_test, y_dpredict))

Mean Absolute Error: 0.5332001304956565
Mean Squared Error: 0.5558915986952442
Root Mean Squared Error: 0.7455813830127763
R2 Score: 0.575787706032451


A Decision Tree Regressor works like a flowchart by asking a series of simple "yes or no" questions about your data features. As it splits the dataset into smaller groups, the final leaves calculate the average house price for each specific group.

In [217]:
from sklearn.ensemble import RandomForestRegressor

# Random Forest handles non-linear relationships natively
rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(x_train_scaled, y_train)

y_pred_rf = rf_model.predict(x_test_scaled)

score = r2_score(y_test, y_pred_rf)
print(f"Random Forest R2 Score: {score:.4f}")


Random Forest R2 Score: 0.8053


A Random Forest Regressor combines multiple individual decision trees to make a single, highly accurate prediction. It trains each tree on a random sample of the data and averages all their final results together to cancel out individual errors.This algorithm is perfect for California housing because combining many trees prevents the model from overfitting or memorizing noise in the data.

In [218]:
from sklearn.ensemble import GradientBoostingRegressor

gb_model = GradientBoostingRegressor(n_estimators=100, random_state=42)
gb_model.fit(x_train_scaled, y_train)

GradientBoostingRegressor(random_state=42)

In [219]:
y_pred_gb = gb_model.predict(x_test_scaled)
y_pred_gb

array([0.50518761, 1.09334601, 4.24570956, ..., 4.68181295, 0.85329537,
       1.96275219])

In [220]:
score1 = r2_score(y_test, y_pred_gb)
print(f"Gradient Boosting Regressor R2 Score: {score1:.4f}")

Gradient Boosting Regressor R2 Score: 0.7756


A Gradient Boosting Regressor builds a team of decision trees sequentially meaning one after the other. Each new tree is specifically trained to correct the exact prediction mistakes made by the previous trees.
This method is excellent for California housing because it focuses heavily on hard-to-predict neighborhoods and complex price patterns. By slowly minimizing errors step-by-step, it achieves some of the highest accuracy possible for this dataset.

3. Model Evaluation and Comparison (2 marks):
● Evaluate the performance of each algorithm using the following metrics:

*   Mean Squared Error (MSE)
*   Mean Absolute Error (MAE)
*   R-squared Score (R2)

● Compare the results of all models and identify: The best-performing algorithm with justification and the worst-performing algorithm with reasoning.

The Linear regression and Decision tree regressor give excatly same accuracy.
MSE of both models is .55, MAE of both models is .53, R2 of both models is .57 and RMSE of both models is .74.

But using ensemble model Random forest we got R2 value .80 and for Gradient Boosting Regressor we got R2 value of .77

Among these models Random Forest is the best accurate with 80 per cent accuracy.